it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,262 @@
+    }
+  }
+
+  private void checkHasInstanceAndDatabase() throws IllegalStateException {
+    if (!hasInstance) {
+      throw new IllegalStateException("There is no instance for manager to perform operation on.");
+    }
+    if (!hasDatabase) {
+      throw new IllegalStateException("There is no database for manager to perform operation on");
+    }
+  }
+
+  /** Creates an instance and database as well if there are none at the time of method call. */
+  @Override
+  public synchronized void createTable(String statement) throws IllegalStateException {
+    checkIsUsable();
+    maybeCreateInstance();
+    maybeCreateDatabase();
+
+    LOG.info("Creating table in database {} using statement '{}'.", databaseId, statement);
+    try {
+      databaseAdminClient
+          .updateDatabaseDdl(
+              instanceId, databaseId, ImmutableList.of(statement), /* operationId= */ null)
+          .get();
+      LOG.info("Successfully created table in database {}.", databaseId);
+    } catch (ExecutionException | InterruptedException | SpannerException e) {
+      throw new SpannerResourceManagerException("Failed to create table.", e);
+    }
+  }
+
+  @Override
+  public synchronized void write(Mutation tableRecord) throws IllegalStateException {
+    write(ImmutableList.of(tableRecord));
+  }
+
+  @Override
+  public synchronized void write(Iterable<Mutation> tableRecords) throws IllegalStateException {
+    checkIsUsable();
+    checkHasInstanceAndDatabase();
+
+    LOG.info("Sending {} mutations to {}.{}", Iterables.size(tableRecords), instanceId, databaseId);
+    try {
+      DatabaseClient databaseClient =
+          spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+      databaseClient.write(tableRecords);
+      LOG.info("Successfully sent mutations to {}.{}", instanceId, databaseId);
+    } catch (SpannerException e) {
+      throw new SpannerResourceManagerException("Failed to write mutations.", e);
+    }
+  }
+
+  @Override
+  public synchronized ImmutableList<Struct> readTableRecords(String tableId, String... columnNames)
+      throws IllegalStateException {
+    return readTableRecords(tableId, ImmutableList.copyOf(columnNames));
+  }
+
+  @Override
+  public synchronized ImmutableList<Struct> readTableRecords(
+      String tableId, Iterable<String> columnNames) throws IllegalStateException {
+    checkIsUsable();
+    checkHasInstanceAndDatabase();
+
+    LOG.info(
+        "Loading columns {} from {}.{}.{}",
+        Iterables.toString(columnNames),
+        instanceId,
+        databaseId,
+        tableId);
+    DatabaseClient databaseClient =
+        spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+
+    try (ResultSet resultSet =
+        databaseClient.singleUse().read(tableId, KeySet.all(), columnNames)) {
+      ImmutableList.Builder<Struct> tableRecordsBuilder = ImmutableList.builder();
+      ImmutableList<Struct> tableRecords;
```

Comment:
I meant below to have: `ImmutableList<Struct> tableRecords = tableRecordsBuilder.build();`

This variable serves no purpose until then.